### ISSUE TITLE
RandomCrop is slow

### ISSUE THREAD

**Issue by vmoens** [kornia/kornia#2175]
### Describe the bug

I was hoping to use batched transforms of Kornia to speed up my training but I realized that RandomCrop is as slow as other libs that execute the transforms serially.

## Context

Executing transforms in a vectorized way can dramatically speed up dataloading, + it's easier to compile.

## Observations:

I executed the following code (I don't need padding or anything fancy, just randomly crop a rectangle of the same size on all of the images of the batch):

```python
import timeit
from kornia.augmentation import RandomCrop as RandomCropKornia
from torch import nn
import torch


class RandomCrop(nn.Module):
    def __init__(self, w, h):
        super(RandomCrop, self).__init__()
        self.w = w
        self.h = h

    def forward(self, x):
        batch = x.shape[:-3]
        index0 = torch.randint(x.shape[-2] - self.h, (*batch, 1), device=x.device)
        index0 = index0 + torch.arange(self.h, device=x.device)
        index0 = (
            index0.unsqueeze(1).unsqueeze(-1).expand(*batch, 3, self.h, x.shape[-1])
        )
        index1 = torch.randint(x.shape[-1] - self.w, (*batch, 1), device=x.device)
        index1 = index1 + torch.arange(self.w, device=x.device)
        index1 = index1.unsqueeze(1).unsqueeze(-2).expand(*batch, 3, self.h, self.w)
        return x.gather(-2, index0).gather(-1, index1)

size = (64, 65)

data = torch.randn(128, 3, 244, 244)


rc_kornia = RandomCropKornia(size)
rc_nn = RandomCrop(*size)

print(timeit.repeat("rc_nn(x)", globals={"rc_nn": rc_nn, "x": data}, number=1_000))
print(timeit.repeat("rc_kornia(x)", globals={"rc_kornia": rc_kornia, "x": data}, number=1_000))

```

On my McBook 2020, I get the following speeds:
```
[4.4548966850000005, 4.295491041, 4.372050861, 4.512908287999998, 4.598690725000001]
[26.677106356, 26.578449804, 27.665404369, 26.978094082, 26.892204961000004]
```
where my batch transform is 5x faster.

## Profiling

I did a bit of profiling. Here are the results (sorry if it is very verbose but I think it's useful). I commented my own random crop.

  <details>
    <summary>Code</summary>

```
Timer unit: 1e-06 s

Total time: 45.9877 s
File: /Users/vmoens/Repos/kornia/kornia/augmentation/_2d/geometric/crop.py
Function: precrop_padding at line 136

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   136                                               @profile
   137                                               def precrop_padding(
   138                                                   self, input: Tensor, padding: Optional[List[int]] = None, flags: Optional[Dict[str, Any]] = None
   139                                               ) -> Tensor:
   140      5000       4539.0      0.9      0.0          flags = self.flags if flags is None else flags
   141      5000       2434.0      0.5      0.0          if padding is None:
   142                                                       padding = self.compute_padding(input.shape)
   143                                           
   144      5000   45976462.0   9195.3    100.0          input = pad(input, padding, value=flags["fill"], mode=flags["padding_mode"])
   145                                           
   146      5000       4250.0      0.8      0.0          return input

Total time: 3.89228 s
File: /Users/vmoens/Repos/kornia/kornia/augmentation/_2d/geometric/crop.py
Function: compute_transformation at line 148

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   148                                               @profile
   149                                               def compute_transformation(self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any]) -> Tensor:
   150      5000       5931.0      1.2      0.2          if flags["cropping_mode"] in ("resample", "slice"):
   151      5000    3883790.0    776.8     99.8              transform: Tensor = get_perspective_transform(params["src"].to(input), params["dst"].to(input))
   152      5000       2554.0      0.5      0.1              return transform
   153                                                   raise NotImplementedError(f"Not supported type: {flags['cropping_mode']}.")

Total time: 88.6386 s
File: /Users/vmoens/Repos/kornia/kornia/augmentation/_2d/geometric/crop.py
Function: apply_transform at line 172

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   172                                               @profile
   173                                               def apply_transform(
   174                                                   self, input: Tensor, params: Dict[str, Tensor], flags: Dict[str, Any], transform: Optional[Tensor] = None
   175                                               ) -> Tensor:
   176      5000       3764.0      0.8      0.0          padding_size: Optional[List[int]] = None
   177      5000       7079.0      1.4      0.0          if "padding_size" in params and isinstance(params["padding_size"], Tensor):
   178      5000    2391265.0    478.3      2.7              padding_size = params["padding_size"].unique(dim=0).cpu().squeeze().numpy().tolist()
   179      5000   46098108.0   9219.6     52.0          input = self.precrop_padding(input, padding_size, flags)
   180                                           
   181      5000       4931.0      1.0      0.0          flags = self.flags if flags is None else flags
   182      5000       6018.0      1.2      0.0          if flags["cropping_mode"] == "resample":  # uses bilinear interpolation to crop
   183                                                       transform = cast(Tensor, transform)
   184                                                       # Fit the arg to F.pad
   185                                                       if flags['padding_mode'] == "constant":
   186                                                           padding_mode = "zeros"
   187                                                       elif flags['padding_mode'] == "replicate":
   188                                                           padding_mode = "border"
   189                                                       elif flags['padding_mode'] == "reflect":
   190                                                           padding_mode = "reflection"
   191                                                       else:
   192                                                           padding_mode = flags['padding_mode']
   193                                           
   194                                                       return crop_by_transform_mat(
   195                                                           input,
   196                                                           transform,
   197                                                           flags["size"],
   198                                                           mode=flags["resample"].name.lower(),
   199                                                           padding_mode=padding_mode,
   200                                                           align_corners=flags["align_corners"],
   201                                                       )
   202      5000       3114.0      0.6      0.0          if flags["cropping_mode"] == "slice":  # uses advanced slicing to crop
   203      5000   40124343.0   8024.9     45.3              return crop_by_indices(input, params["src"], flags["size"])
   204                                                   raise NotImplementedError(f"Not supported type: {flags['cropping_mode']}.")

Total time: 6.05657 s
File: /Users/vmoens/Repos/kornia/kornia/augmentation/_2d/geometric/crop.py
Function: forward_parameters at line 293

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   293                                               @profile
   294                                               def forward_parameters(self, batch_shape) -> Dict[str, Tensor]:
   295      5000      33012.0      6.6      0.5          input_pad = self.compute_padding(batch_shape)
   296      5000      10213.0      2.0      0.2          batch_shape_new = (
   297      5000      15638.0      3.1      0.3              *batch_shape[:2],
   298      5000       6378.0      1.3      0.1              batch_shape[2] + input_pad[2] + input_pad[3],  # original height + top + bottom padding
   299      5000       3347.0      0.7      0.1              batch_shape[3] + input_pad[0] + input_pad[1],  # original width + left + right padding
   300                                                   )
   301      5000     263688.0     52.7      4.4          padding_size = tensor(tuple(input_pad), dtype=torch.long).expand(batch_shape[0], -1)
   302      5000    5716605.0   1143.3     94.4          _params = super().forward_parameters(batch_shape_new)
   303      5000       5882.0      1.2      0.1          _params.update({"padding_size": padding_size})
   304      5000       1804.0      0.4      0.0          return _params

Total time: 37.746 s
File: /Users/vmoens/Repos/kornia/kornia/geometry/transform/crop2d.py
Function: crop_by_indices at line 286

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   286                                           @profile
   287                                           def crop_by_indices(
   288                                               input_tensor: Tensor,
   289                                               src_box: Tensor,
   290                                               size: Optional[Tuple[int, int]] = None,
   291                                               interpolation: str = 'bilinear',
   292                                               align_corners: Optional[bool] = None,
   293                                               antialias: bool = False,
   294                                               shape_compensation: str = "resize",
   295                                           ) -> Tensor:
   296                                               """Crop tensors with naive indices.
   297                                           
   298                                               Args:
   299                                                   input: the 2D image tensor with shape (B, C, H, W).
   300                                                   src_box: a tensor with shape (B, 4, 2) containing the coordinates of the bounding boxes
   301                                                       to be extracted. The tensor must have the shape of Bx4x2, where each box is defined in the clockwise
   302                                                       order: top-left, top-right, bottom-right and bottom-left. The coordinates must be in x, y order.
   303                                                   size: output size. An auto resize or pad will be performed according to ``shape_compensation``
   304                                                       if the cropped slice sizes are not exactly align `size`.
   305                                                       If None, will auto-infer from src_box.
   306                                                   interpolation:  algorithm used for upsampling: ``'nearest'`` | ``'linear'`` | ``'bilinear'`` |
   307                                                       'bicubic' | 'trilinear' | 'area'.
   308                                                   align_corners: interpolation flag.
   309                                                   antialias: if True, then image will be filtered with Gaussian before downscaling.
   310                                                       No effect for upscaling.
   311                                                   shape_compensation: if the cropped slice sizes are not exactly align `size`, the image can either be padded
   312                                                       or resized.
   313                                               """
   314      5000      24229.0      4.8      0.1      B, C, _, _ = input_tensor.shape
   315      5000     144041.0     28.8      0.4      src = as_tensor(src_box, device=input_tensor.device, dtype=torch.long)
   316      5000     148991.0     29.8      0.4      x1 = src[:, 0, 0]
   317      5000     156467.0     31.3      0.4      x2 = src[:, 1, 0] + 1
   318      5000      50024.0     10.0      0.1      y1 = src[:, 0, 1]
   319      5000      82372.0     16.5      0.2      y2 = src[:, 3, 1] + 1
   320                                           
   321      5000       5020.0      1.0      0.0      if (
   322      5000     355366.0     71.1      0.9          len(x1.unique(sorted=False))
   323      5000     185368.0     37.1      0.5          == len(x2.unique(sorted=False))
   324      5000     199609.0     39.9      0.5          == len(y1.unique(sorted=False))
   325       355      11613.0     32.7      0.0          == len(y2.unique(sorted=False))
   326       355        134.0      0.4      0.0          == 1
   327                                               ):
   328                                                   out = input_tensor[..., int(y1[0]) : int(y2[0]), int(x1[0]) : int(x2[0])]
   329                                                   if size is not None and out.shape[-2:] != size:
   330                                                       return resize(
   331                                                           out, size, interpolation=interpolation, align_corners=align_corners, side="short", antialias=antialias
   332                                                       )
   333                                           
   334      5000       3508.0      0.7      0.0      if size is None:
   335                                                   h, w = infer_bbox_shape(src)
   336                                                   size = h.unique(sorted=False), w.unique(sorted=False)
   337      5000      71092.0     14.2      0.2      out = torch.empty(B, C, *size, device=input_tensor.device, dtype=input_tensor.dtype)
   338                                               # Find out the cropped shapes that need to be resized.
   339    640000    1210751.0      1.9      3.2      for i, _ in enumerate(out):
   340    640000   18099511.0     28.3     48.0          _out = input_tensor[i : i + 1, :, int(y1[i]) : int(y2[i]), int(x1[i]) : int(x2[i])]
   341    640000    1152231.0      1.8      3.1          if _out.shape[-2:] != size:
   342                                                       if shape_compensation == "resize":
   343                                                           out[i] = resize(
   344                                                               _out,
   345                                                               size,
   346                                                               interpolation=interpolation,
   347                                                               align_corners=align_corners,
   348                                                               side="short",
   349                                                               antialias=antialias,
   350                                                           )
   351                                                       else:
   352                                                           out[i] = pad(_out, [0, size[1] - _out.shape[-1], 0, size[0] - _out.shape[-2]])
   353                                                   else:
   354    640000   15842930.0     24.8     42.0              out[i] = _out
   355      5000       2694.0      0.5      0.0      return out
```
  </details>

I am surprised that padding takes so much time when I did not ask for padding at all. I'm not a CV guy so please forgive me if I'm missing the point here.

### Reproduction steps

```bash
See above
```

### Expected behavior

Batched transforms should be super fast. I noticed that there's a for loop in the `crop_by_indices` function. I think this could be vectorized (see my example) in some cases. The implementation should (1) detect such cases and (2) take this shortcut if possible.

### Environment

```shell
Kornia installed from clone.
Latest torch nightly.
McBook 2020 (not M1). No GPU but happy to reprod on CUDA.
```

**Comment by edgarriba** [comment]
@vmoens thanks for the feedback. I guess this could be an upgrade of the existing `crop_by_indices` in here
https://github.com/kornia/kornia/blob/master/kornia/geometry/transform/crop2d.py#L287

/cc @shijianjian @johnnv1

**Comment by shijianjian** [comment]
Interesting that the padding is a pretty slow operation. After update to:

```python
    def precrop_padding(
        self, input: Tensor, padding: Optional[List[int]] = None, flags: Optional[Dict[str, Any]] = None
    ) -> Tensor:
        flags = self.flags if flags is None else flags
        if padding is None:
            padding = self.compute_padding(input.shape)

        # vvvvvv The update
        if any(padding):
            input = pad(input, padding, value=flags["fill"], mode=flags["padding_mode"])

        return input
```

We can get:
```bash 
[5.921756925, 6.9425626739999995, 6.568994523999999, 7.860189029000001, 7.356724565]
[4.842821216000004, 5.157504515999996, 5.479773302000005, 5.157385547000004, 4.902601369999999]
```

**Comment by edgarriba** [comment]
@shijianjian did you compare against same implementation provided by @vmoens  ?

**Comment by shijianjian** [comment]
It is the comparison between his and mine.

**Comment by shijianjian** [comment]
closed by #2179
